File: src/frontend/src/types/quality.ts
export type EntityKind = 'data_domain' | 'data_product' | 'data_contract' | 'asset';

export type QualityDimension =
  | 'accuracy'
  | 'completeness'
  | 'conformity'
  | 'consistency'
  | 'coverage'
  | 'timeliness'
  | 'uniqueness'
  | 'other';

export type QualitySource =
  | 'manual'
  | 'dbt'
  | 'dqx'
  | 'great_expectations'
  | 'soda'
  | 'external';

export const QUALITY_DIMENSIONS: QualityDimension[] = [
  'accuracy', 'completeness', 'conformity', 'consistency',
  'coverage', 'timeliness', 'uniqueness', 'other',
];

export const QUALITY_SOURCES: QualitySource[] = [
  'manual', 'dbt', 'dqx', 'great_expectations', 'soda', 'external',
];

export interface QualityItem {
  id: string;
  entity_id: string;
  entity_type: EntityKind;
  title?: string | null;
  description?: string | null;
  dimension: QualityDimension;
  source: QualitySource;
  score_percent: number;
  checks_passed?: number | null;
  checks_total?: number | null;
  measured_at: string;
  created_by?: string | null;
  updated_by?: string | null;
  created_at: string;
  updated_at: string;
}

export interface QualityItemCreate {
  entity_id: string;
  entity_type: EntityKind;
  title?: string | null;
  description?: string | null;
  dimension: QualityDimension;
  source?: QualitySource;
  score_percent: number;
  checks_passed?: number | null;
  checks_total?: number | null;
  measured_at?: string | null;
}

export interface QualityItemUpdate extends Partial<Omit<QualityItemCreate, 'entity_id' | 'entity_type'>> {}

export interface QualitySummary {
  overall_score_percent: number;
  items_count: number;
  by_dimension: Record<string, number>;
  by_source: Record<string, number>;
  measured_at?: string | null;
}

export const dimensionColors: Record<QualityDimension, string> = {
  accuracy: '#3b82f6',
  completeness: '#22c55e',
  conformity: '#8b5cf6',
  consistency: '#f59e0b',
  coverage: '#06b6d4',
  timeliness: '#ec4899',
  uniqueness: '#f97316',
  other: '#64748b',
};

export function scoreColor(pct: number): string {
  if (pct >= 90) return '#22c55e';
  if (pct >= 70) return '#f59e0b';
  return '#ef4444';
}

File: src/backend/src/models/quality.py
from __future__ import annotations
from typing import Optional, Dict
from uuid import UUID
from datetime import datetime

from pydantic import BaseModel, Field


QUALITY_DIMENSIONS = ("accuracy", "completeness", "conformity", "consistency", "coverage", "timeliness", "uniqueness", "other")
QUALITY_SOURCES = ("manual", "dbt", "dqx", "great_expectations", "soda", "external")

_DIM_PATTERN = r"^(" + "|".join(QUALITY_DIMENSIONS) + r")$"
_SRC_PATTERN = r"^(" + "|".join(QUALITY_SOURCES) + r")$"
_ENTITY_PATTERN = r"^(data_domain|data_product|data_contract|asset)$"


class QualityItemBase(BaseModel):
    entity_id: str
    entity_type: str = Field(..., pattern=_ENTITY_PATTERN)
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = Field(None, max_length=2000)
    dimension: str = Field(..., pattern=_DIM_PATTERN)
    source: str = Field("manual", pattern=_SRC_PATTERN)
    score_percent: float = Field(..., ge=0.0, le=100.0)
    checks_passed: Optional[int] = Field(None, ge=0)
    checks_total: Optional[int] = Field(None, ge=0)
    measured_at: Optional[datetime] = None


class QualityItemCreate(QualityItemBase):
    pass


class QualityItemUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=255)
    description: Optional[str] = Field(None, max_length=2000)
    dimension: Optional[str] = Field(None, pattern=_DIM_PATTERN)
    source: Optional[str] = Field(None, pattern=_SRC_PATTERN)
    score_percent: Optional[float] = Field(None, ge=0.0, le=100.0)
    checks_passed: Optional[int] = Field(None, ge=0)
    checks_total: Optional[int] = Field(None, ge=0)
    measured_at: Optional[datetime] = None


class QualityItem(QualityItemBase):
    id: UUID
    created_by: Optional[str] = None
    updated_by: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    model_config = {
        "from_attributes": True
    }


class QualitySummary(BaseModel):
    overall_score_percent: float
    items_count: int
    by_dimension: Dict[str, float]
    by_source: Dict[str, float]
    measured_at: Optional[datetime] = None
